### PR TITLE
11092 paperclip => active storage

### PR DESCRIPTION
**Pull request by cooperka**
Ready for high-level feedback.

### Remaining TODOs before it's production-ready

- [x] Make sure response download filenames include the response shortcode like before (will require DB migration for legacy data)
- [x] Confirm that ODK respects media redirects
- [x] Add authorization constraints + ability specs for new ActiveStorage download endpoints
- [x] Get active_storage:copy_from_paperclip task fully working (see 2 rake TODOs)
- [x] Test end-to-end migration on real data (only with 'local'? convert cloud servers to local for faster migrations?)
- [x] Confirm we can do a local-only migration easily by temporarily copying cloud attachments
- [x] Partially or fully migrate servers to .env (because config/storage.rb must depend on ENV, not Settings)
- [x] Purge paperclip & fog dependencies
- [x] Make sure we document clear upgrade instructions

### Deployment instructions

To run with real data, either locally or on production:

1. Initial verifications for cloud servers:
    - Make sure there's nothing in the bucket other than `uploads/` (if so note it so you don't delete it accidentally when cleaning up)
    - Make sure you've granted the database server IP access to the bucket contents to view/copy
    - Make a backup of the bucket in case you need to purge duplicates and restart
1. Check out main, import real data, and smoke test
    - Upload Paperclip images as desired for testing
1. If the server has a LOT of cloud files and you want to speed up the conversion migration:
    - Copy the S3 files into local `uploads/` folder on the database server: `cd elmo/shared;
aws s3 cp --recursive s3://nemo-bucket-name .`
    - Temporarily set `NEMO_STORAGE_TYPE=local` in `.env.production`
    - Check out/deploy the commit BEFORE `Add active_storage:copy_s3 task from HB`
    - `rails db:migrate` to insert into new active storage metadata tables
    - Revert back to `NEMO_STORAGE_TYPE=cloud`
1. Check out/deploy the commit BEFORE `[Breaking]`
1. Increase the thread pool in `database.yml` to 100 to allow parallelization if using cloud storage (set `NUM_THREADS` lower if things fail)
1. `rails db:migrate` to create new ActiveStorage tables and add associations
    - This migration will also copy files for `local` servers, or prepare some data for `cloud` servers to use in the next step
    - This change is backwards-compatible
1. Check out/deploy latest commit
1. `rails db:migrate` to copy files on S3 into new locations and update the ActiveStorage tables
    - This migration is only needed for `cloud` servers (no-op on `local` servers)
    - This change is backwards-compatible
1. Smoke test
    - The old Paperclip images you attached in the first step should now load and render via ActiveStorage
1. Cleanup: Decrease the thread pool back to 5, re-block bucket access
1. (optional, breaking, later) Delete the old attachment blobs (both cloud and local versions) and Paperclip columns (including legacy_url) from the schema

### Resources

- Similar PR: https://github.com/sassafrastech/heartmob/pull/45
- Migration guide: https://github.com/thoughtbot/paperclip/blob/master/MIGRATING.md
- ActiveStorage docs: https://guides.rubyonrails.org/v6.0/active_storage_overview.html

### CODE REVIEW

File: spec/models/itemsets_form_attachment_spec.rb [hunk 1]
```diff
@@ -29,37 +29,37 @@
...
+        allow(itemsets_attachment).to receive(:file_contents).and_return("foo")
+        expect(itemsets_attachment.md5).to eq("acbd18db4cc2f85cedef654fccc4a4d8") # This is md5 of "foo"
       end
     end
   end
 
   describe "empty?" do
     before do
-      allow(ifa).to receive(:decorated_form).and_return(double(needs_external_csv?: needs_external_csv))
+      allow(itemsets_attachment).to receive(:decorated_form).and_return(double(needs_external_csv?: needs_external_csv))
```
[the remark this answers is not in the document]
Comment (smoyth) [reply spec/models/itemsets_form_attachment_spec.rb:62]: Curious if your editor makes it clear when you're past 110...?

File: spec/models/media/object_spec.rb [hunk 2]
```diff
@@ -31,6 +31,6 @@
   let(:media_file) { create(:media_image) }
 
   it "has attachment" do
-    expect(media_file).to have_attached_file(:item)
+    expect(media_file.item.attached?).to be_truthy
```
Comment (smoyth) [inline spec/models/media/object_spec.rb:34]: I prefer `be(true)` if you know it's going to be a boolean.

File: spec/models/results/response_json_generator_spec.rb [hunk 3]
```diff
@@ -44,7 +46,10 @@
     end
 
     it "produces correct json" do
-      is_expected.to match_json(prepare_response_json_expectation("basic.json", image_id: [image.id]))
+      path = Rails.application.routes.url_helpers.rails_blob_url(image.item, disposition: "attachment",
+                                                                             only_path: true)
+      "#{Results::ResponseJsonGenerator::BASE_URL_PLACEHOLDER}#{path}"
```
[the remark this answers is not in the document]
Comment (smoyth) [reply spec/models/results/response_json_generator_spec.rb:51]: What is going on here?

**Review by smoyth — COMMENTED** [review]
Phew! Looks great! What a lot of work! Kudos! Lots of good attention to detail and cleanups. I left a few comments. Please also tend to Hound issues. Thanks!

File: config/environments/production.rb [hunk 4]
```diff
@@ -55,8 +55,9 @@
   # config.action_dispatch.x_sendfile_header = 'X-Sendfile' # for Apache
   # config.action_dispatch.x_sendfile_header = 'X-Accel-Redirect' # for NGINX
 
-  # Store uploaded files on the local file system (see config/storage.yml for options).
-  config.active_storage.service = :local
+  # Where to store uploaded files (see config/storage.yml for options).
+  storage_type = ENV["NEMO_STORAGE_TYPE"].presence || Settings.paperclip&.storage
+  config.active_storage.service = storage_type == "cloud" ? :amazon : :local
```
[the remark this answers is not in the document]
Comment (cooperka) [reply config/environments/production.rb:60]: I propose we handle this afterward -- I've added it to my list of next steps not blocking deploy if that works for you.

**Comment by cooperka** [comment]
I've cherry-picked in your commits @smoyth and then force-pushed with updated migration logic immediately before and after `[breaking]` commit. Tested and verified on staging. I'll have plenty more to say at standup 🚀 

Just a few little comments remaining + TODO for maintaining filenames. I'll start on that tomorrow morning after standup unless you feel like going for it earlier.

File: config/environments/production.rb [hunk 5]
```diff
@@ -55,8 +55,9 @@
   # config.action_dispatch.x_sendfile_header = 'X-Sendfile' # for Apache
   # config.action_dispatch.x_sendfile_header = 'X-Accel-Redirect' # for NGINX
 
-  # Store uploaded files on the local file system (see config/storage.yml for options).
-  config.active_storage.service = :local
+  # Where to store uploaded files (see config/storage.yml for options).
+  storage_type = ENV["NEMO_STORAGE_TYPE"].presence || Settings.paperclip&.storage
```
[the remark this answers is not in the document]
Comment (cooperka) [reply config/environments/production.rb]: I removed `Settings.paperclip` here after removing the Paperclip dependency entirely 👍 Will wait on .env for future PR

**Comment by cooperka** [comment]
I versioned it v12.0 because BOOYAH

**Comment by cooperka** [comment]
Specs are flappers. This seems ready to merge, then follow up with #804

**Pull request merged**